WT-15767 Fix a race condition between checkpoint and stepping up

pmacko86 opened the pull request:
Hold a checkpoint lock during step up / step down, because otherwise changing the leader state `conn->layered_table_manager.leader` concurrently with a checkpoint would end up leaving the system in an inconsistent state.

pmacko86 committed:
WT-15767 Fix a race condition between checkpoint and stepping up

pmacko86 committed:
Add an error code check

mariammojid commented on src/conn/conn_layered.c:
```diff
@@ -1129,6 +1129,73 @@ __disagg_restart_checkpoint(WT_SESSION_IMPL *session)
+    F_SET(conn, WT_CONN_RECONFIGURING_STEP_UP);
+
+    /* Step up to the leader mode. */
+    conn->layered_table_manager.leader = true;
+    WT_STAT_CONN_SET(session, disagg_role_leader, 1);
```
Would it be better to move these two steps after draining the ingest stable? That way we know for certain these were set after the necessary steps succeeded. If there is no way we can be left in that state then please disregard.

pmacko86 replied:
No, actually both of these tasks must happen first. I added comments explaining that in [9914ee0](https://github.com/wiredtiger/wiredtiger/pull/12644/commits/9914ee0ee3cb54ed634ff27bbb8d1d9618dcc1f0).

mariammojid replied:
Thank you for adding that comment!

mariammojid commented on src/conn/conn_layered.c:
```diff
@@ -1129,6 +1129,73 @@ __disagg_restart_checkpoint(WT_SESSION_IMPL *session)
     return (ret);
 }
 
+/*
```
Could you please create a python test that ensures this race condition is addressed?

pmacko86 replied:
Oh yeah—I got a bit lazy last night. I added a test in [1ba1b6e](https://github.com/wiredtiger/wiredtiger/pull/12644/commits/1ba1b6ee2dab8179c9e3e8a4891eeb721dea1ad1).

mariammojid replied:
Awesome :) TY

pmacko86 committed:
Merge branch 'develop' into wt-15767-step-up-fix

pmacko86 committed:
Added a test

pmacko86 committed:
Added more detailed comments

mariammojid reviewed (approved):
LGTM! NP - would be nice to have some verbose logs.

ershov commented:
LGTM!
It'd be helpful if a high level description was included in the test. What is the test testing and how?

pmacko86 committed:
Renamed test_layered61.py -> test_layered62.py

pmacko86 committed:
Added extra comments & logs

pmacko86 committed:
Merge branch 'develop' into wt-15767-step-up-fix

pmacko86 commented:
Thanks for the reviews!

> NP - would be nice to have some verbose logs.

I added logs for stepping up and stepping down. The individual steps within that already have customized error messages.

>  It'd be helpful if a high level description was included in the test. What is the test testing and how?

I just added a two-paragraph comment to the test. It would be nice to add such comments to other tests too...
